File: script/blog.swift
import Foundation

func getAllFilePath(_ dirPath: String) -> [String] {    
    guard let array = try? FileManager.default.contentsOfDirectory(atPath: dirPath) else {
        return []
    }
    var filePaths = [String]()
    for fileName in array {
        var isDir: ObjCBool = true
        let fullPath = "\(dirPath)/\(fileName)"
        if FileManager.default.fileExists(atPath: fullPath, isDirectory: &isDir) {
            if isDir.boolValue {
                filePaths.append(contentsOf: getAllFilePath(fullPath))
            } else {
                filePaths.append(fullPath)
            }
        }
    }
    return filePaths
}

struct Blog: Codable {
    var title: String
    var fileName: String
    var date: String
}

if let idx = CommandLine.arguments.firstIndex(of: "-p") {
    let path = CommandLine.arguments[idx + 1]
    let paths = getAllFilePath(path).filter({ $0.hasSuffix(".md")})
    var blogs = [Blog]()
    for subPath in paths.sorted().reversed() {
        let file = (try? String.init(contentsOfFile: subPath)) ?? ""
        let lines = file.split(separator: "\n")
        let name = lines[0].dropFirst()
        let time = lines[1].dropFirst()
        let fileName = subPath.split(separator: "/").last ?? ""

        let blog = Blog(title: String(name), fileName: String(fileName), date: String(time))
        blogs.append(blog)
    }
    let encoder = JSONEncoder()
    encoder.outputFormatting = .prettyPrinted
    let data = try! encoder.encode(blogs)
    let json = String(data: data, encoding: .utf8)!
    print(json)
}


// swift ./script/blog.swift -p $(pwd)/blog > ./terminal/blog.json

File: scripts/blog.mjs
#!/usr/bin/env zx

// import "zx/globals";


const files = fs.readdirSync("../blog").filter((file) => file.endsWith(".md"))  
let blogs = []

for (const file of files) {
    const content = await fs.readFile(`../blog/${file}`, "utf8");
    const lines = content.split("\n");
    const title = lines[0].replace("# ", "");
    const date = lines[1].replace("> ", "");   
    blogs.push({
        title: title,
        date: date,
        fileName: file,
    })
}  

blogs = blogs.sort((a, b) => {
    return new Date(b.date) - new Date(a.date);
})
const fileContent = JSON.stringify(blogs, null, 2);
fs.writeFileSync("../database/blogs.json", fileContent);